File: scripts/bootstrap.py
#!/usr/bin/env python3

'''
Suppository Bootstrap Script

Personalizes a new project created from the Suppository GitHub template.

Usage:
    python scripts/bootstrap.py
'''
from pathlib import Path
import os
import shutil
import subprocess
import sys


ROOT = Path(__file__).parent.parent

SKIP_DIRS = {
    ".git",
    "node_modules",
    ".next",
}


PLACEHOLDERS = {
    "{{PROJECT_NAME}}": None,
    "{{PACKAGE_NAME}}": None,
    "{{DESCRIPTION}}": None,
    "{{AUTHOR}}": None,
    "{{GITHUB_USERNAME}}": None,
}


def run(command: list[str], optional=False):
    '''Run a shell command.'''
    print(f"\n> {' '.join(command)}")

    try:
        subprocess.run(
            command,
            cwd=ROOT,
            check=True,
        )
    except subprocess.CalledProcessError:
        if optional:
            print("Skipped")
        else:
            raise


def ask(prompt, default=None):
    '''Prompt user.'''
    if default:
        value = input(
            f"{prompt} [{default}]: "
        ).strip()

        return value or default

    return input(
        f"{prompt}: "
    ).strip()


def collect_information():
    print('''
=================================
🚀 Suppository Project Bootstrap
=================================
''')

    project = ask(
        "Project name"
    )

    package = ask(
        "Package name",
        project.lower()
        .replace(" ", "-")
    )

    PLACEHOLDERS["{{PROJECT_NAME}}"] = project
    PLACEHOLDERS["{{PACKAGE_NAME}}"] = package

    PLACEHOLDERS["{{DESCRIPTION}}"] = ask(
        "Description"
    )

    PLACEHOLDERS["{{AUTHOR}}"] = ask(
        "Author"
    )

    PLACEHOLDERS["{{GITHUB_USERNAME}}"] = ask(
        "GitHub username"
    )


def iter_files():
    for path in ROOT.rglob("*"):
        if not path.is_file():
            continue

        if any(
            part in SKIP_DIRS
            for part in path.parts
        ):
            continue
        yield path


def replace_placeholders():
    print("\nReplacing placeholders...\n")

    for file in iter_files():
        try:
            contents = file.read_text(
                encoding="utf-8"
            )
        except UnicodeDecodeError:
            continue

        original = contents

        for key, value in PLACEHOLDERS.items():
            if value:
                contents = contents.replace(
                    key,
                    value
                )
        if contents != original:
            file.write_text(
                contents,
                encoding="utf-8"
            )

            print(f"✓ {file}")


def create_env():
    example = ROOT / ".env.example"
    env = ROOT / ".env"

    if env.exists():
        print("✓ .env already exists")
        return

    if example.exists():
        shutil.copy(
            example,
            env
        )

        print("✓ Created .env")


def rename_package():
    old = ROOT / "src" / "package"

    package = PLACEHOLDERS[
        "{{PACKAGE_NAME}}"
    ]

    if old.exists():
        old.rename(
            old.parent / package
        )

        print(
            f"✓ Renamed package to {package}"
        )


def initialize_git():
    if (ROOT / ".git").exists():
        return

    answer = ask(
        "Initialize git repository? (y/n)",
        "y"
    )

    if answer.lower() == "y":
        run(
            ["git", "init"]
        )

        run(
            [
                "git",
                "add",
                "."
            ]
        )

        run(
            [
                "git",
                "commit",
                "-m",
                "Initial commit"
            ],
            optional=True
        )


def install_dependencies():
    answer = ask(
        "Install dependencies? (y/n)",
        "y"
    )

    if answer.lower() == "y":
        run(
            [
                "pnpm",
                "install"
            ]
        )


def setup_database():
    answer = ask(
        "Initialize database? (y/n)",
        "y"
    )

    if answer.lower() == "y":
        run(
            [
                "pnpm",
                "prisma",
                "generate"
            ]
        )

        run(
            [
                "pnpm",
                "prisma",
                "migrate",
                "dev",
                "--name",
                "init"
            ]
        )


def main():
    collect_information()
    replace_placeholders()
    create_env()
    rename_package()
    initialize_git()
    install_dependencies()
    setup_database()

    print('''
=================================

🎉 Bootstrap complete!

Next steps:

    pnpm dev

Your application is ready.

=================================
''')


if __name__ == "__main__":
    main()

File: scripts/bootstrap.ts
#!/usr/bin/env tsx
/**
 * Suppository Bootstrap Script
 *
 * Personalizes a new project created from the Suppository template.
 *
 * Run:
 *   pnpm bootstrap
 */
import fs from "node:fs/promises";
import path from "node:path";
import { existsSync } from "node:fs";
import readline from "node:readline/promises";
import { stdin as input, stdout as output } from "node:process";
import { execSync } from "node:child_process";


const ROOT = path.resolve(
  import.meta.dirname,
  ".."
);

const SKIP_DIRS = new Set([
  ".git",
  "node_modules",
  ".next",
]);

const placeholders: Record<string, string> = {
  "{{PROJECT_NAME}}": "",
  "{{PACKAGE_NAME}}": "",
  "{{DESCRIPTION}}": "",
  "{{AUTHOR}}": "",
  "{{GITHUB_USERNAME}}": "",
};

const rl = readline.createInterface({
  input,
  output,
});


function log(message: string) {
  console.log(message);
}


async function ask(
  question: string,
  defaultValue?: string
) {
  const suffix = defaultValue
    ? ` [${defaultValue}]`
    : "";

  const answer = await rl.question(
    `${question}${suffix}: `
  );
  return answer.trim() || defaultValue || "";
}


function run(
  command: string,
  optional = false
) {
  log(`\n> ${command}`);

  try {
    execSync(command, {
      cwd: ROOT,
      stdio: "inherit",
    });
  } catch (error) {
    if (!optional) {
      throw error;
    }

    log("Skipped");
  }
}


async function collectInfo() {
  console.log(`
=================================
🚀 Suppository Bootstrap
=================================
`);
  placeholders["{{PROJECT_NAME}}"] =
    await ask("Project name");

  placeholders["{{PACKAGE_NAME}}"] =
    await ask(
      "Package name",
      placeholders["{{PROJECT_NAME}}"]
        .toLowerCase()
        .replaceAll(" ", "-")
    );

  placeholders["{{DESCRIPTION}}"] =
    await ask("Description");

  placeholders["{{AUTHOR}}"] =
    await ask("Author");

  placeholders["{{GITHUB_USERNAME}}"] =
    await ask("GitHub username");
}


async function getFiles(
  directory: string
): Promise<string[]> {
  const files: string[] = [];

  const entries = await fs.readdir(
    directory,
    {
      withFileTypes: true,
    }
  );

  for (const entry of entries) {
    const fullPath = path.join(
      directory,
      entry.name
    );

    if (
      entry.isDirectory()
    ) {
      if (
        SKIP_DIRS.has(entry.name)
      ) {
        continue;
      }

      files.push(
        ...(await getFiles(fullPath))
      );
    } else {
      files.push(fullPath);
    }
  }
  return files;
}


async function replacePlaceholders() {
  log("\nReplacing placeholders...\n");
  const files = await getFiles(ROOT);

  for (const file of files) {
    try {
      let content =
        await fs.readFile(
          file,
          "utf8"
        );

      const original = content;

      for (
        const [key,value]
        of Object.entries(placeholders)
      ) {
        if (value) {
          content =
            content.replaceAll(
              key,
              value
            );
        }
      }

      if (content !== original) {
        await fs.writeFile(
          file,
          content,
          "utf8"
        );

        log(
          `✓ ${path.relative(ROOT,file)}`
        );
      }
    } catch {
      // Ignore binary files
    }
  }
}


async function createEnv() {
  const envExample =
    path.join(ROOT, ".env.example");

  const env =
    path.join(ROOT, ".env");

  if (
    existsSync(env)
  ) {
    log("✓ .env already exists");
    return;
  }

  if (
    existsSync(envExample)
  ) {
    await fs.copyFile(
      envExample,
      env
    );

    log("✓ Created .env");
  }
}


function initializeGit() {
  if (
    existsSync(
      path.join(ROOT, ".git")
    )
  ) {
    return;
  }

  run(
    "git init"
  );

  run(
    "git add ."
  );

  run(
    'git commit -m "Initial commit"',
    true
  );
}


function renamePackage() {
  const oldPath =
    path.join(
      ROOT,
      "src",
      "package"
    );

  if (
    !existsSync(oldPath)
  ) {
    return;
  }

  const newPath =
    path.join(
      ROOT,
      "src",
      placeholders["{{PACKAGE_NAME}}"]
    );

  execSync(
    `mv "${oldPath}" "${newPath}"`
  );

  log(
    `✓ Renamed package folder`
  );
}


async function installDependencies() {
  const answer =
    await ask(
      "Install dependencies? (y/n)",
      "y"
    );

  if (
    answer.toLowerCase() === "y"
  ) {
    run(
      "pnpm install"
    );
  }
}


async function setupDatabase() {
  const answer =
    await ask(
      "Initialize database? (y/n)",
      "y"
    );

  if (
    answer.toLowerCase() === "y"
  ) {
    run(
      "pnpm prisma generate"
    );

    run(
      "pnpm prisma migrate dev --name init"
    );
  }
}


async function main() {
  await collectInfo();
  await replacePlaceholders();
  await createEnv();
  renamePackage();
  initializeGit();
  await installDependencies();
  await setupDatabase();

  console.log(`
=================================

🎉 Bootstrap complete!

Start development:

    pnpm dev

=================================
`);

  rl.close();
}


main()
  .catch((error) => {
    console.error(error);
    rl.close();
    process.exit(1);
  });
